### ISSUE TITLE
Movie Recommendation System

### ISSUE THREAD

**Issue by Pavi0512** [Pavi0512/Movie_recommation#1]
Movie Recommendation System

Build a simple recommendation system that suggests movies based on user ratings or genres.
Implement collaborative filtering or content-based filtering techniques.

**Comment by Pavi0512** [comment]
"""
neumf_rec.py
A full example of Neural Collaborative Filtering (NeuMF) using MovieLens 100K (or any ratings CSV).
Requirements:
    pip install numpy pandas tensorflow scikit-learn tqdm
Run:
    python neumf_rec.py
"""

import os
import zipfile
import numpy as np
import pandas as pd
from tqdm import tqdm
import tensorflow as tf
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
import urllib.request

# -------------------------
# 1) Download MovieLens 100K (if not present)
# -------------------------
ML_100K_URL = "http://files.grouplens.org/datasets/movielens/ml-100k.zip"
DATA_DIR = "ml-100k"

if not os.path.exists(DATA_DIR):
    print("Downloading MovieLens 100K...")
    zip_path = "ml-100k.zip"
    urllib.request.urlretrieve(ML_100K_URL, zip_path)
    with zipfile.ZipFile(zip_path, "r") as z:
        z.extractall(".")
    os.remove(zip_path)
    print("Downloaded and extracted.")

# MovieLens u.data format: user id | item id | rating | timestamp
ratings_path = os.path.join(DATA_DIR, "u.data")
ratings = pd.read_csv(ratings_path, sep="\t", header=None, names=["user_id", "item_id", "rating", "timestamp"])
ratings = ratings[["user_id", "item_id", "rating"]]

# Optional: load movie titles for nicer output
movies = pd.read_csv(os.path.join(DATA_DIR, "u.item"), sep="|", header=None, encoding="latin-1",
                     names=["item_id", "title"] + list(range(22)), usecols=[0,1])

# -------------------------
# 2) Preprocess: create mappings, train/test split, negative sampling for implicit ranking tasks
# -------------------------
# We will treat this as implicit feedback: any rating >=4 -> positive (1); else negative (0)
ratings["label"] = (ratings["rating"] >= 4).astype(int)

# Encode user and item IDs to contiguous ints starting at 0 for embedding layers
user_encoder = LabelEncoder()
item_encoder = LabelEncoder()
ratings["user_idx"] = user_encoder.fit_transform(ratings["user_id"])
ratings["item_idx"] = item_encoder.fit_transform(ratings["item_id"])
num_users = ratings["user_idx"].nunique()
num_items = ratings["item_idx"].nunique()

print(f"Users: {num_users}, Items: {num_items}, Interactions: {len(ratings)}")

# Create train/test split by leaving one positive interaction per user for test (leave-one-out)
def leave_one_out(df):
    df_sorted = df.sort_values("user_idx")
    train_list = []
    test_list = []
    for u, group in df_sorted.groupby("user_idx"):
        pos = group[group["label"] == 1]
        neg = group[group["label"] == 0]
        if len(pos) == 0:
            # if no positive interactions, just put all into train
            train_list.append(group)
            continue
        # take 1 positive for test randomly
        test_pos = pos.sample(1, random_state=42)
        train_pos = pos.drop(test_pos.index)
        test_group = pd.concat([test_pos, neg.sample(min(len(neg), 99), random_state=42)])  # keep some negatives to test
        train_group = pd.concat([train_pos, neg.drop(test_group.index, errors="ignore")])
        train_list.append(train_group)
        test_list.append(test_group)
    train_df = pd.concat(train_list).reset_index(drop=True)
    test_df = pd.concat(test_list).reset_index(drop=True) if len(test_list)>0 else pd.DataFrame(columns=df.columns)
    return train_df, test_df

train_df, test_df = leave_one_out(ratings)

# For training NeuMF with binary cross-entropy we want positive and negative pairs.
# We'll perform negative sampling for each positive instance.
def negative_sampling(train, num_items, neg_ratio=4):
    users, items, labels = [], [], []
    user_pos = train[train["label"]==1].groupby("user_idx")["item_idx"].apply(set).to_dict()
    rng = np.random.default_rng(42)
    for _, row in train[train["label"]==1].iterrows():
        u = row["user_idx"]
        i_pos = row["item_idx"]
        users.append(u); items.append(i_pos); labels.append(1)
        # sample negatives
        n = 0
        while n < neg_ratio:
            j = rng.integers(0, num_items)
            if j not in user_pos.get(u, set()):
                users.append(u); items.append(j); labels.append(0)
                n += 1
    return np.array(users), np.array(items), np.array(labels)

train_users, train_items, train_labels = negative_sampling(train_df, num_items, neg_ratio=4)
print("Training examples (after negative sampling):", len(train_labels))

# For validation/test, we'll evaluate ranking metrics using the left-out positives and sampled negatives in test_df
test_df = test_df.reset_index(drop=True)

# -------------------------
# 3) Build NeuMF model (GMF + MLP combined)
# -------------------------
embedding_dim = 32
mlp_layers = [64, 32, 16]

user_input = tf.keras.layers.Input(shape=(), dtype=tf.int32, name="user")
item_input = tf.keras.layers.Input(shape=(), dtype=tf.int32, name="item")

# GMF part (element-wise product of user/item embeddings)
gmf_user_emb = tf.keras.layers.Embedding(num_users, embedding_dim, name="gmf_user_emb")(user_input)
gmf_item_emb = tf.keras.layers.Embedding(num_items, embedding_dim, name="gmf_item_emb")(item_input)
gmf_user_vec = tf.keras.layers.Flatten()(gmf_user_emb)
gmf_item_vec = tf.keras.layers.Flatten()(gmf_item_emb)
gmf_out = tf.keras.layers.Multiply()([gmf_user_vec, gmf_item_vec])

# MLP part
mlp_user_emb = tf.keras.layers.Embedding(num_users, embedding_dim, name="mlp_user_emb")(user_input)
mlp_item_emb = tf.keras.layers.Embedding(num_items, embedding_dim, name="mlp_item_emb")(item_input)
mlp_user_vec = tf.keras.layers.Flatten()(mlp_user_emb)
mlp_item_vec = tf.keras.layers.Flatten()(mlp_item_emb)
mlp_concat = tf.keras.layers.Concatenate()([mlp_user_vec, mlp_item_vec])

x = mlp_concat
for i, layer_size in enumerate(mlp_layers):
    x = tf.keras.layers.Dense(layer_size, activation="relu", name=f"mlp_dense_{i}")(x)
mlp_out = x

# Concatenate GMF and MLP
concat = tf.keras.layers.Concatenate()([gmf_out, mlp_out])
concat = tf.keras.layers.Dense(64, activation="relu")(concat)
output = tf.keras.layers.Dense(1, activation="sigmoid")(concat)

model = tf.keras.Model(inputs=[user_input, item_input], outputs=output)
model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
              loss="binary_crossentropy",
              metrics=[tf.keras.metrics.AUC(name="auc")])

model.summary()

# -------------------------
# 4) Train
# -------------------------
batch_size = 1024
epochs = 6

history = model.fit(
    x=[train_users, train_items],
    y=train_labels,
    batch_size=batch_size,
    epochs=epochs,
    validation_split=0.1,
    verbose=1
)

# -------------------------
# 5) Evaluation: Hit Ratio@K and NDCG@K
# -------------------------
def evaluate_model(model, test_df, k=10):
    hits = []
    ndcgs = []
    # build mapping: for each user, list of candidate items (test positive + negatives)
    for u, group in test_df.groupby("user_idx"):
        # positive item(s) in this test group
        pos_items = group[group["label"]==1]["item_idx"].tolist()
        if len(pos_items)==0: 
            continue
        pos_item = pos_items[0]  # left-out positive
        candidate_items = group["item_idx"].tolist()  # includes pos + negatives
        users = np.array([u]*len(candidate_items))
        preds = model.predict([users, np.array(candidate_items)], batch_size=len(candidate_items), verbose=0).flatten()
        # rank items by pred score
        idx_sorted = np.argsort(-preds)
        ranked_items = np.array(candidate_items)[idx_sorted]
        # Hit@K
        hit = 1.0 if pos_item in ranked_items[:k] else 0.0
        hits.append(hit)
        # NDCG@K
        if pos_item in ranked_items[:k]:
            rank = np.where(ranked_items[:k]==pos_item)[0][0] + 1  # rank starts at 1
            ndcg = 1.0 / np.log2(rank + 1)
        else:
            ndcg = 0.0
        ndcgs.append(ndcg)
    hr = np.mean(hits) if hits else 0.0
    mean_ndcg = np.mean(ndcgs) if ndcgs else 0.0
    return hr, mean_ndcg

hr10, ndcg10 = evaluate_model(model, test_df, k=10)
print(f"HitRate@10: {hr10:.4f}, NDCG@10: {ndcg10:.4f}")

# -------------------------
# 6) Recommendation function: recommend top-N items for a user
# -------------------------
def recommend_top_n(model, user_raw_id, n=10):
    # map raw user id to idx; if not seen, return popular items
    try:
        u_idx = user_encoder.transform([user_raw_id])[0]
    except Exception:
        print("User not found in training data — returning most popular items.")
        popular = ratings.groupby("item_idx").size().sort_values(ascending=False).index[:n]
        return movies[movies["item_id"].isin(item_encoder.inverse_transform(popular))]["title"].tolist()
    # score all items
    all_items = np.arange(num_items)
    users = np.array([u_idx]*num_items)
    scores = model.predict([users, all_items], batch_size=1024, verbose=0).flatten()
    top_idx = np.argsort(-scores)[:n]
    top_item_idx = all_items[top_idx]
    # convert item_idx back to raw item_id and titles
    raw_item_ids = item_encoder.inverse_transform(top_item_idx)
    titles = movies[movies["item_id"].isin(raw_item_ids)]["title"].tolist()
    return titles

# Example: recommend for user 1 (raw ID)
example_user = ratings["user_id"].unique()[0]
print("Example user raw id:", example_user)
print("Top 10 recommendations (titles):")
print(recommend_top_n(model, example_user, n=10))
